### PR TITLE
Fix uninitialized memory and realpath behaviour in ModelicaInternal_fullPathName

### LINKED ISSUE #3003: ModelicaInternal_fullPathName truncates trailing '/' on Linux
The function ModelicaInternal_fullPathName truncates trailing '/' on Linux, but not on Windows.
The truncating behavior in the Linux realpath needs to be compensated for, by checking for existing trailing '/' in the input argument and then add this later to the resulting fullName if needed.

### PR DESCRIPTION
**Pull request by beutlich**
close #3003

### CODE REVIEW

**Comment by beutlich** [comment]
> 
> 
> I don't like the "len = strlen(name)-1;" due to the potential that it is an empty string.
> Instead of investigating that in detail I would prefer:
> 
> ```
> ... len =strlen(name);
> if (len>0 && ('/' ==name[len-1] || '\\'==name[len-1])) {
> ```

Right. Will update the PR.

**Comment by beutlich** [comment]
For the sake of completeness: [Here](https://gist.github.com/beutlich/88e7662045e4197d8e9add8c5756b44f) is my test case.

File: Modelica/Resources/C-Sources/ModelicaInternal.c [hunk 1]
```diff
@@ -588,9 +592,18 @@ _Ret_z_ const char* ModelicaInternal_fullPathName(_In_z_ const char* name) {
...
     }
-    fullName = ModelicaAllocateString(strlen(tempName));
+    fullName = ModelicaAllocateString(strlen(tempName) + 1);
     strcpy(fullName, tempName);
     ModelicaConvertToUnixDirectorySeparator(fullName);
+#if (_BSD_SOURCE || _XOPEN_SOURCE >= 500 || _XOPEN_SOURCE && _XOPEN_SOURCE_EXTENDED || _POSIX_VERSION >= 200112L)
+    {
+        /* In case of realpath: Retain trailing slash to match _fullpath behaviour */
+        size_t len = strlen(name);
+        if (len > 0 && ('/' == name[len - 1] || '\\' == name[len - 1])) {
```
Comment (sjoelund) [inline Modelica/Resources/C-Sources/ModelicaInternal.c]: Looking for '\\' here will give the wrong behaviour. You are changing the path from a file to a directory (remember: '\\' on POSIX systems is a file name, not a path separator).
The only problem is if you want MINGW to be handled differently. In that case those defines might require `&& !defined(MINGW)` and using `_fullpath` instead (if that's what you want to mimic, why not call it directly?).

Comment (beutlich) [reply Modelica/Resources/C-Sources/ModelicaInternal.c]: I tested Cygwin only I need to admit. Can you update my example above with the test case you have in mind, please! Thanks.

Comment (beutlich) [reply Modelica/Resources/C-Sources/ModelicaInternal.c]: Hm, I am rather confused with your comment. Whereas Cygwin uses the realpath implementation, MinGW uses the _fullpath implementation (and is not affected by this PR). Tested with

```
tbeu@LAPTOP-RUH9UABA /cygdrive/C/temp
$ gcc --version
gcc (GCC) 7.4.0
Copyright (C) 2017 Free Software Foundation, Inc.
Dies ist freie Software; die Kopierbedingungen stehen in den Quellen. Es
gibt KEINE Garantie; auch nicht für MARKTGÄNGIGKEIT oder FÜR SPEZIELLE ZWECKE.
```

and

```
C:\temp>gcc --version
gcc (tdm64-1) 5.1.0
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Comment (sjoelund) [reply Modelica/Resources/C-Sources/ModelicaInternal.c]: Oh, true... You don't include unistd.h for MINGW... And for Cygwin you don't use \ as path separator. So on that line the check for '\\' at the end of the string should simply be removed as we're dealing with Unix systems.

Comment (beutlich) [reply Modelica/Resources/C-Sources/ModelicaInternal.c]: Right. Will update PR now,

**Pull request merged**